draft/mson-parameters-headers.md

```diff
@@ -0,0 +1,216 @@
+- [Motivation](#motivation)
+- [Implementation](#implementation)
+- [Backwards Compatibility](#backwards-compatibility)
+
+## Abstract
+
+This RFC proposes using full [MSON][] syntax for description of URI parameters
+and HTTP headers.
+
+## Motivation
+
+Using MSON syntax for description of URI parameters and HTTP Headers will
+greatly help to unify the API Blueprint syntax and introduce reusability of
+URI parameters and HTTP headers. It will also greatly introduce the reusability
+of other data structures.
+
+In addition it will address many issues previously tracked on API Blueprint
+GitHub repository.
+
+### URI Parameters Issues Tracked
+
+This RFC addresses following issues:
+
+1. [Usage of data structures in Parameters](https://github.com/apiaryio/api-blueprint/issues/250)
+1. [Global query URI parameter](https://github.com/apiaryio/api-blueprint/issues/24)
+1. [Description of enumeration elements](https://github.com/apiaryio/api-blueprint/issues/119)
+1. [Array in parameters](https://github.com/apiaryio/api-blueprint/issues/240)
+1. [Minimum/Maximum values for URI Parameters](https://github.com/apiaryio/api-blueprint/issues/211)
+
+    NOTE: This is going to be addressed once MSON introduces validation.
+
+### HTTP Headers Issues Tracked
+
+This RFC addresses following issues:
+
+1. [Parametric Headers Description](https://github.com/apiaryio/api-blueprint/issues/26)
+1. [Referenced HTTP headers](https://github.com/apiaryio/api-blueprint/issues/17)
+1. [Support for multiline syntax header](https://github.com/apiaryio/api-blueprint/issues/60)
+
+In addition this issue is solved by this RFC:
+
+1. [Reuse headers and attributes](https://github.com/apiaryio/api-blueprint/issues/259)
+
+## Implementation
+
+### Parameters
+
+The implementation should assume the existing syntax for API Blueprint
+[URI parameters section][] is, in fact, an [MSON type declaration][].
+
+The type being declared inherits from a *Parameters base type* unless specified
+otherwise. If another base type is specified it must inherit from the
+*Parameters base type*.
+
+The *Parameter base type* may the [API Description Namespace][]
+[`Href Variables`][] type.
+
+The *Parameters base type* should be available as a type for use within any
+[data structures section][].
+
+#### Example
+
+```apib
+# Resource [/resource/{p}{?q}]
++ Parameters
+    + p
+    + q
+```
+
+Assuming `Parameters` as the name of the *Parameters base type*:
+
+```apib
+# Resource [/resource/{p}{?q}]
++ Parameters (My Parameters)
+
+# Data Structures
+## My Parameters (Parameters)
++ p
++ q
+```
+
+### Headers
+
+The implementation of this RFC should introduce the use of MSON syntax in the
+[Headers section][]. Similarly to Parameters section this Headers section should
+be considered as [MSON type declaration][].
+
+The type being declared inherits from a *Headers base type* unless specified
+otherwise. If another base type is specified it must inherit from the
+*Headers base type*.
```

Comment:
Another argument against it is that it introduces a notion of global base types that are domain specific. They wouldn't fit in MSON, so you'd have to add them as a base type in maybe a MSON Apiary layer which seems like it would introduce a new moving part which makes the whole system just a little bit harder to understand (reading the MSON doc would not suffice to understand where the global base types come from).